lib/cloudant.js

```diff
@@ -604,29 +604,39 @@ Cloudant.prototype.replaceOrCreate = function replaceOrCreate(
   var self = this;
   var idName = self.idName(model);
   var mo = self.selectModel(model);
-
   var id = data[idName];
-  if (id) {
-    self.getCurrentRevision(model, id, function(err, rev) {
-      if (err) return cb(err);
-      if (rev) data._rev = rev;
-      self._insert(model, data, function(err) {
+
+  // Check if data exist first to avoid
+  // running getCurrentRevision() again an nonexistent instance
+  var where = {}; where[idName] = id;
+  var dataExist;
+
+  self.count(model, where, {}, function(err, cnt) {
+    if (err) return cb(err);
+    dataExist = cnt === 1 ? true : false;
+
+    if (id && dataExist === true) {
+      self.getCurrentRevision(model, id, function(err, rev) {
         if (err) return cb(err);
-        mo.db.get(id, function(err, doc) {
+        if (rev) data._rev = rev;
+        self._insert(model, data, function(err) {
           if (err) return cb(err);
-          cb(err, self.fromDB(model, mo, doc), {isNewInstance: !rev});
+          mo.db.get(id, function(err, doc) {
+            if (err) return cb(err);
+            cb(err, self.fromDB(model, mo, doc), {isNewInstance: !rev});
+          });
         });
       });
-    });
-  } else {
-    self.create(model, data, options, function(err, id) {
-      if (err) return cb(err);
-      mo.db.get(id, function(err, doc) {
+    } else {
+      self.create(model, data, options, function(err, id) {
         if (err) return cb(err);
-        cb(err, self.fromDB(model, mo, doc), {isNewInstance: true});
+        mo.db.get(id, function(err, doc) {
+          if (err) return cb(err);
+          cb(err, self.fromDB(model, mo, doc), {isNewInstance: true});
+        });
       });
-    });
-  }
+    };
+  });
```

Comment:
I'm not that much familiar with cloudant; is this the only way to do it in cloudant? Isn't there any atomic method to use? I trust your judgment in this regard...